source/extensions/formatter/xfcc_value/xfcc_value.cc

```diff
@@ -0,0 +1,215 @@
+    return std::string(raw_value);
+  }
+
+  // Unescape double quotes.
+  // Note: this only handles escaped double quotes, not other escape sequences.
+  std::string unescaped;
+  unescaped.reserve(raw_value.size());
+  for (size_t i = 0; i < raw_value.size(); ++i) {
+    if (raw_value[i] == '\\' && i + 1 < raw_value.size() && raw_value[i + 1] == '"') {
+      unescaped.push_back('"');
+      ++i;
+    } else {
+      unescaped.push_back(raw_value[i]);
+    }
+  }
+  return unescaped;
+}
+
+// Handles a single XFCC element (semicolon-separated key/value pairs).
+absl::optional<std::string> parseElementForKey(absl::string_view element,
+                                               absl::string_view target) {
+
+  // Scan key-value pairs in this element (by semicolon not in quotes).
+  bool in_quotes = false;
+  size_t start = 0;
+  const size_t element_size = element.size();
+  for (size_t i = 0; i <= element_size; ++i) {
+    // Check for end of key-value pair.
+    if (i == element_size || element[i] == ';') {
+      // If not in quotes then we found the end of a key-value pair.
+      if (!in_quotes) {
+        auto value = parseKeyValuePair(element.substr(start, i - start), target);
+        if (value.has_value()) {
+          return value;
+        }
+        start = i + 1;
+      }
+      continue;
+    }
+
+    // Switch quote state if we encounter a quote character.
+    if (element[i] == '"') {
+      if (i == 0 || element[i - 1] != '\\') {
+        in_quotes = !in_quotes;
+      }
+    }
+  }
+
+  // Note, we should never encounter unmatched quotes here because if there is
+  // an unmatched quote, it should be handled in the parseValueFromXfccByKey()
+  // and will not enter this function.
+  ASSERT(!in_quotes);
+  return absl::nullopt;
+}
+
+// Extracts the key from the XFCC header.
+absl::StatusOr<std::string> parseValueFromXfccByKey(const Http::RequestHeaderMap& headers,
+                                                    absl::string_view target) {
+  absl::string_view value = headers.getForwardedClientCertValue();
+  if (value.empty()) {
+    return absl::InvalidArgumentError("XFCC header is not present");
+  }
+
+  // Scan elements in the XFCC header (by comma not in quotes).
+  bool in_quotes = false;
+  size_t start = 0;
+  const size_t value_size = value.size();
+  for (size_t i = 0; i <= value_size; ++i) {
+    // Check for end of element.
+    if (i == value_size || value[i] == ',') {
+      // If not in quotes then we found the end of an element.
+      if (!in_quotes) {
+        auto result = parseElementForKey(value.substr(start, i - start), target);
+        if (result.has_value()) {
+          return result.value();
+        }
+        start = i + 1;
+      }
+      continue;
+    }
+
+    // Switch quote state if we encounter a quote character.
+    if (value[i] == '"') {
+      if (i == 0 || value[i - 1] != '\\') {
+        in_quotes = !in_quotes;
+      }
+    }
+  }
```

Comment:
Will this work for some string like `by="a\\";uri="spiffe://good"`? if `i` is at the second `"`, and `value[i-1] != '\\'`, then this will lead to some invalid status?

Reply:
Nope. This is not supported and no plan to support it for now because it will make the parsing more complex.
As the comment said, only slash + `"` will be handled and except this case, `slash` to `slash` cannot be handled.
Because there is no constraints about this in the format definition of xfcc header.